viskores/exec/SplineEvaluateRectilinearGrid.h

```diff
@@ -0,0 +1,259 @@
+    for (viskores::Id kk = 0; kk < 4; ++kk)
+    {
+      viskores::Id k = viskores::Clamp(iw - 1 + kk, 0, this->NumZ - 1);
+      for (viskores::Id jj = 0; jj < 4; ++jj)
+      {
+        viskores::Id j = viskores::Clamp(iv - 1 + jj, 0, this->NumY - 1);
+        for (viskores::Id ii = 0; ii < 4; ++ii)
+        {
+          viskores::Id i = viskores::Clamp(iu - 1 + ii, 0, this->NumX - 1);
+          auto pIndex = (kk * 4 + jj) * 4 + ii;
+          auto dIndex = (k * this->NumY + j) * this->NumX + i;
+          P[pIndex] = this->Field.Get(dIndex);
+        }
+      }
+    }
+
+    // interpolate in X for each (kk,jj) → Cbuf[16]
+    viskores::FloatDefault Cbuf[4 * 4];
+    for (int kk = 0; kk < 4; ++kk)
+      for (int jj = 0; jj < 4; ++jj)
+      {
+        auto x0 = this->AxisPortals[0].Get(iu - 1);
+        auto x1 = this->AxisPortals[0].Get(iu);
+        auto x2 = this->AxisPortals[0].Get(iu + 1);
+        auto x3 = this->AxisPortals[0].Get(iu + 2);
+        auto p0 = P[(kk * 4 + jj) * 4 + 0];
+        auto p1 = P[(kk * 4 + jj) * 4 + 1];
+        auto p2 = P[(kk * 4 + jj) * 4 + 2];
+        auto p3 = P[(kk * 4 + jj) * 4 + 3];
+        Cbuf[kk * 4 + jj] = this->CubicInterpolateNonUniform(x0, x1, x2, x3, p0, p1, p2, p3, x);
+      }
+
+    // interpolate in Y → D[4]
+    viskores::FloatDefault D[4];
+    for (int kk = 0; kk < 4; ++kk)
+    {
+      auto y0 = this->AxisPortals[1].Get(iv - 1);
+      auto y1 = this->AxisPortals[1].Get(iv);
+      auto y2 = this->AxisPortals[1].Get(iv + 1);
+      auto y3 = this->AxisPortals[1].Get(iv + 2);
+      D[kk] = this->CubicInterpolateNonUniform(
+        y0, y1, y2, y3, Cbuf[kk * 4 + 0], Cbuf[kk * 4 + 1], Cbuf[kk * 4 + 2], Cbuf[kk * 4 + 3], y);
+    }
+
+    // interpolate in Z
+    auto z0 = this->AxisPortals[2].Get(iw - 1);
+    auto z1 = this->AxisPortals[2].Get(iw);
+    auto z2 = this->AxisPortals[2].Get(iw + 1);
+    auto z3 = this->AxisPortals[2].Get(iw + 2);
+    value = this->CubicInterpolateNonUniform(z0, z1, z2, z3, D[0], D[1], D[2], D[3], z);
+
+    return viskores::ErrorCode::Success;
+  }
+
+private:
+  VISKORES_EXEC viskores::Id FindIndex(const AxisPortalType& axis,
+                                       const viskores::Id& N,
+                                       viskores::FloatDefault val) const
+  {
+    // 1) Binary search for the largest index i with coords[i] <= val
+    viskores::Id left = 0;
+    viskores::Id right = N - 1;
+    while (left <= right)
+    {
+      viskores::Id mid = left + (right - left) / 2;
+      if (axis.Get(mid) <= val)
+      {
+        // mid is still ≤ val, so it might be our i
+        left = mid + 1;
+      }
+      else
+      {
+        // coords[mid] > val, so the index we want is below mid
+        right = mid - 1;
+      }
+    }
+    // when loop ends, `right` is the last index where coords[right] <= val
+    viskores::Id i = right;
+
+    // 2) Clamp i into [1, N-3]
+    if (i < 1)
+      i = 1;
+    else if (i > N - 3)
+      i = N - 3;
```

Comment:
Why not use `viskores::Clamp`? I just spent a good part of this morning obsessing about the function you added.

```suggestion
    viskores::Clamp(i, 1, N - 3);
```